File: shims/node/go-secrets-preload.js
// go-secrets Node.js preload shim for Windows
const net = require('net');

const PIPE_NAME = '\\\\.\\pipe\\go-secrets-daemon';

const AUTH_TOKEN = process.env.__SECRETS_AUTH_TOKEN;
if (!AUTH_TOKEN) {
  console.error('[go-secrets] ERROR: __SECRETS_AUTH_TOKEN not set. Run via: secrets env run -- node ...');
  process.exit(1);
}

delete process.env.__SECRETS_AUTH_TOKEN;

const secretsCache = {};

const helperScript = `
const net = require('net');
const AUTH_TOKEN = process.argv[2];
const PIPE_NAME = '\\\\\\\\.\\\\pipe\\\\go-secrets-daemon';

const client = net.connect(PIPE_NAME, () => {
  const request = JSON.stringify({
    type: 'get_all_secrets',
    auth_token: AUTH_TOKEN
  });
  client.end(request + '\\n');
});

let responseData = '';
client.on('data', (data) => {
  responseData += data.toString();
});

client.on('end', () => {
  try {
    const response = JSON.parse(responseData);
    if (response.success && response.secrets) {
      console.log(JSON.stringify(response.secrets));
    } else {
      console.error('ERROR: ' + (response.error || 'Failed'));
      process.exit(1);
    }
  } catch (err) {
    console.error('ERROR: ' + err.message);
    process.exit(1);
  }
});

client.on('error', (err) => {
  console.error('ERROR: ' + err.message);
  process.exit(1);
});
`;

const { execFileSync } = require('child_process');
const fs = require('fs');
const os = require('os');
const path = require('path');

const helperPath = path.join(os.tmpdir(), 'go-secrets-helper.js');
fs.writeFileSync(helperPath, helperScript);

try {
  console.log('[go-secrets] Loading secrets from daemon...');
  const result = execFileSync('node', [helperPath, AUTH_TOKEN], {
    encoding: 'utf8',
    timeout: 5000
  });
  const secrets = JSON.parse(result.trim());
  Object.assign(secretsCache, secrets);
  console.log(`[go-secrets] Loaded ${Object.keys(secrets).length} secret(s)`);
} catch (err) {
  console.error(`[go-secrets] WARNING: Failed to load secrets: ${err.message}`);
}

try { fs.unlinkSync(helperPath); } catch (e) {}

const originalEnv = process.env;

const envProxy = new Proxy(originalEnv, {
  get(target, prop) {
    if (prop in target) {
      return target[prop];
    }
    if (prop in secretsCache) {
      return secretsCache[prop];
    }
    return undefined;
  },

  has(target, prop) {
    return prop in target || prop in secretsCache;
  },

  ownKeys(target) {
    return [...Reflect.ownKeys(target), ...Object.keys(secretsCache)];
  },

  getOwnPropertyDescriptor(target, prop) {
    if (prop in target) {
      return Object.getOwnPropertyDescriptor(target, prop);
    }
    
    if (prop in secretsCache) {
      return {
        enumerable: true,
        configurable: true,
        writable: false,
        value: secretsCache[prop]
      };
    }

    return undefined;
  }
});

Object.defineProperty(process, 'env', {
  get: () => envProxy,
  set: () => {
    throw new Error('Cannot replace process.env');
  },
  enumerable: true,
  configurable: false
});

File: shims/python/go_secrets_preload.py
# go-secrets Python preload shim for Windows
import os
import json
import sys

PIPE_NAME = r'\\.\pipe\go-secrets-daemon'

AUTH_TOKEN = os.environ.get('__SECRETS_AUTH_TOKEN')
if not AUTH_TOKEN:
    print('[go-secrets] ERROR: __SECRETS_AUTH_TOKEN not set. Run via: secrets env run -- python ...', file=sys.stderr)
    sys.exit(1)

del os.environ['__SECRETS_AUTH_TOKEN']

secrets_cache = {}

def fetch_all_secrets():
    try:
        import win32file
        
        handle = win32file.CreateFile(
            PIPE_NAME,
            win32file.GENERIC_READ | win32file.GENERIC_WRITE,
            0,
            None,
            win32file.OPEN_EXISTING,
            0,
            None
        )
        
        request = json.dumps({
            'type': 'get_all_secrets',
            'auth_token': AUTH_TOKEN
        }).encode('utf-8') + b'\n'
        
        win32file.WriteFile(handle, request)
        
        result, data = win32file.ReadFile(handle, 65536)
        response = json.loads(data.decode('utf-8'))
        
        win32file.CloseHandle(handle)
        
        if response.get('success') and response.get('secrets'):
            return response['secrets']
        else:
            print(f"[go-secrets] ERROR: {response.get('error', 'Failed to fetch secrets')}", file=sys.stderr)
            return {}
        
    except ImportError:
        print('[go-secrets] ERROR: pywin32 not installed. Install with: pip install pywin32', file=sys.stderr)
        return {}
    except Exception as e:
        print(f'[go-secrets] ERROR: Failed to load secrets: {e}', file=sys.stderr)
        return {}

print('[go-secrets] Loading secrets from daemon...')
secrets_cache = fetch_all_secrets()
print(f'[go-secrets] Loaded {len(secrets_cache)} secret(s)')

_original_environ = dict(os.environ)

class SecretsEnviron(dict):
    def __init__(self):
        super().__init__(_original_environ)
        self.update(secrets_cache)
    
    def __getitem__(self, key):
        if key in _original_environ:
            return _original_environ[key]
        if key in secrets_cache:
            return secrets_cache[key]
        raise KeyError(key)
    
    def get(self, key, default=None):
        try:
            return self[key]
        except KeyError:
            return default
    
    def __contains__(self, key):
        return key in _original_environ or key in secrets_cache
    
    def keys(self):
        return set(_original_environ.keys()) | set(secrets_cache.keys())
    
    def values(self):
        return [self[k] for k in self.keys()]
    
    def items(self):
        return [(k, self[k]) for k in self.keys()]

os.environ = SecretsEnviron()
